colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,336 @@
+                            if self.use_zero:
+                                update = _split(
+                                    input_=update_reshape.view(-1), dim=-1, process_group=self.data_parallel_group
+                                )
+                            else:
+                                update = update_reshape
+                        else:
+                            update_reshape = update.view(-1, grad_shape[1])
+                            grad_reshape = grad.view(-1, grad_shape[1])
+                            exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/dp/tp]
+                            exp_avg_sq_col = state["exp_avg_sq_col"]  # [W]
+                            exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                            exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                            # reduce col
+                            dist.all_reduce(exp_avg_sq_col, group=self.tensor_parallel_group)
+                            exp_avg_sq_col.div_(self.tensor_parallel_size)
+                            # gather row
+                            exp_avg_sq_row_gather = _gather(
+                                input_=exp_avg_sq_row, dim=-1, process_group=self.tensor_parallel_group
+                            )
+                            sq_row_meam = exp_avg_sq_row_gather.mean(dim=-1, keepdim=True)
+                            update_reshape = self._approx_sq_grad_row_parallel(
+                                exp_avg_sq_row, exp_avg_sq_col, sq_row_meam
+                            )
+                            update_reshape.mul_(grad_reshape)
+                            if self.use_zero:
+                                update = update_reshape.view(-1)
+                            else:
+                                update = update_reshape
+                else:
+                    exp_avg_sq = state["exp_avg_sq"]
+                    exp_avg_sq.mul_(beta2t).add_(update, alpha=(1.0 - beta2t))
+                    update = exp_avg_sq.rsqrt().mul_(grad)
+
+                # (Line No.8) RMS
+                # perform a sum on each device
+                update_sum = update.pow(2).sum()
+                num_of_element = update.numel()
+                # reduce sum on tp group if exist
+                if self.tensor_parallel_size > 1 and param_is_dtensor:
+                    dist.all_reduce(update_sum, group=self.tensor_parallel_group)
+                    num_of_element = num_of_element * self.tensor_parallel_size
+                # reduce sum on dp group if exist
+                if self.data_parallel_size > 1 and param_is_dtensor:
+                    dist.all_reduce(update_sum, group=self.data_parallel_group)
+                    num_of_element = num_of_element * self.data_parallel_size
+                # div num of element
+                rms = (update_sum / num_of_element).sqrt()
```

Comment:
Move to `self._rms()`

Reply:
> Move to `self._rms()`


Moved in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/e783599672f9898c86de97d3e1330f4f27b43f49;